From 8ad0f59ef58504d09b8c6c7b6ee90cd367161c3a Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Fri, 2 Jun 2017 17:54:57 +0200
Subject: [PATCH 2/2] RunTest: adjust ISO-8859-1 tests for Cygwin
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

PCRE2 has a test that verifies the "starting code unit" for several
ISO-8859-1 tests. In particular, it wants to verify that the code points
0xAA, 0xB5 and 0xBA in ISO-8859-1 (which translate to ª, µ and º,
respectively) are considered to be word characters, 0xB5 even a
lower-case one.

However, Cygwin does _not_ consider those characters to be word
characters, but considers them _punctuation_ characters:
https://github.com/cygwin/cygwin/blob/cygwin-3.4.2/newlib/libc/ctype/ctype_iso.h#L10-L12
(punctuation characters are marked up with a `_P`, see
https://github.com/cygwin/cygwin/blob/cygwin-3.4.2/newlib/libc/include/ctype.h#L98,
and alphanumerical characters are marked up with `_U` for upper-case,
`_L` for lower-case and `_N` for numerical characters:
https://github.com/cygwin/cygwin/blob/cygwin-3.4.2/newlib/libc/include/ctype.h#L99).

Therefore, these tests will always fail.

Let's work around this by specifically testing whether the mu character
is considered a word character, and if not, edit the expectations
accordingly.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 RunTest | 38 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/RunTest b/RunTest
index ae32f5d..9651592 100755
--- a/RunTest
+++ b/RunTest
@@ -572,6 +572,44 @@ for bmode in "$test8" "$test16" "$test32"; do
             sed "s/fr_FR/$loc/" $testdata/testoutput3A >test3outputA
             sed "s/fr_FR/$loc/" $testdata/testoutput3B >test3outputB
           fi
+          # On Cygwin, the `mu` character (\xaa) in ISO-8859-1 is not
+          # considered to be an alphanumerical character. Likewise
+	  # \xb5 and \xba. Work around PCRE2's expectation that they are.
+          case "$(LC_CTYPE=C \
+                  printf '/\\w/locale=fr_FR\n\xb5\n' |
+                  pcre2test)" in
+          *"No match"*)
+            for f in $infile $outfile $outfile2 $outfile3; do
+              LC_CTYPE=C \
+              sed -e '
+                # remove the three characters from
+                # [A-Za-z\xaa\xb5\xba\xc0-\xd6\xd8-\xf6\xf8-\xff] and from
+                # [a-z\xb5\xdf-\xf6\xf8-\xff]
+                s/a-z\(\\x[ab][0-9a-f]\)*/a-z/
+
+                # remove the three characters from the "Starting code units"
+                # character range
+                /\xaa \xb5 \xba /{
+                  s/\xaa \xb5 \xba //
+                  # rewrap after removing three characters
+                  N;N
+                  s/\(\n *\)\(. . . \)/\2\1/g
+                }
+
+                # test E-circumflex and U-acute which Cygwin considers
+                # alphanumerical characters instead of the ordinal indicators
+                # that Cygwin does not consider to be alphanumerical, both
+                # in the \x... form as well as in the encoded form
+                s/xaa/xca/g' -e 's/xba/xda/g
+                s/\xaa/\xca/g' -e 's/\xba/\xda/g' \
+                <$f >$f-cygwin
+            done
+	    infile=$infile-cygwin
+	    outfile=$outfile-cygwin
+	    outfile2=$outfile2-cygwin
+	    outfile3=$outfile3-cygwin
+            ;;
+          esac
           break
         fi
       fi
